src/DecryptionFailureTracker.ts

```diff
@@ -203,44 +316,63 @@ export class DecryptionFailureTracker {
      * Clear state and stop checking for and tracking failures.
      */
     public stop(): void {
+        if (this.client && this.keysChangedHandler) {
+            this.client.removeListener(CryptoEvent.KeysChanged, this.keysChangedHandler);
+        }
+        this.client = undefined;
+        this.keysChangedHandler = undefined;
+
         if (this.checkInterval) clearInterval(this.checkInterval);
         if (this.trackInterval) clearInterval(this.trackInterval);
 
+        this.userTrustsOwnIdentity = undefined;
         this.failures = new Map();
         this.visibleEvents = new Set();
-        this.visibleFailures = new Map();
-        this.failureCounts = new Map();
+        this.failuresToReport = new Set();
     }
 
     /**
-     * Mark failures that occurred before nowTs - GRACE_PERIOD_MS as failures that should be
-     * tracked. Only mark one failure per event ID.
+     * Mark failures as undecryptable or late. Only mark one failure per event ID.
+     *
      * @param {number} nowTs the timestamp that represents the time now.
      */
     public checkFailures(nowTs: number): void {
-        const failuresGivenGrace: Set<DecryptionFailure> = new Set();
+        const failures: Set<DecryptionFailure> = new Set();
         const failuresNotReady: Map<string, DecryptionFailure> = new Map();
-        for (const [eventId, failure] of this.visibleFailures) {
-            if (nowTs > failure.ts + DecryptionFailureTracker.GRACE_PERIOD_MS) {
-                failuresGivenGrace.add(failure);
+        for (const [eventId, failure] of this.failures) {
+            if (
+                failure.timeToDecryptMillis ||
```

Comment:
I think this does the wrong thing if `timeToDecryptMillis === 0`. Avoid implicit booleanification where possible, it frequently bites you on the ass

Reply:
`timeToDecryptMillis` only ever gets set to `MAXIMUM_LATE_DECRYPTION_PERIOD` or higher.  But it probably is better to be more explicit about the check.